### DISCUSSION TITLE
Pyinstaller loading libraries during runtime

### DISCUSSION THREAD

**Discussion by TheGrandJonsson** [pyinstaller/pyinstaller#6622] [Help]
Hi!
I have a question that I've been searching for an answer to for a while. 
I'm trying to build an application with PyInstaller that's using a package which loads multiple dll files during runtime. 
When I freeze this application with PyInstaller an try to run the app, I get a PyInstallerImportError: Failed to load dynlib/dll. 
I've tried to add the binaries explicitly whit "--add-binary", dumping the libraries in to dist after build, modifying the spec file with the path to the libraries, and changing the path from which the package is loading the binaries. But nothing seems to work. 

I read in some other issue or SO post that PyInstaller has (at least, _had_) some issues with loading ctypes during runtime, so I'm wondering if this could be related to my issue. 
I'm mostly checking to see if someone has some kind of idéa to point me in some direction or if I'll have to give up and search for another solution to my issue. 
I'm using PyInstaller 4.9 btw. 

Thank's tho for an awesome package in all other ways! :)

**Comment by rokm** [comment]
Can you provide a specific example, so we can check what's going on?

`PyInstallerImportError` does come from our ctypes override (which in addition to the original specified location, also does a fall-back check in the application's top-level directory). But that by itself doesn't say much, except that the library either could not be found or could not be loaded for whatever reason (for example, due to a missing dependency).

**Reply by TheGrandJonsson** [reply]
Yes, so the python package i'm using is loading multiple libraries as such: `ctypes.cdll.LoadLibrary(os.path.join(path, additional_lib))` 
The path to those libraires resides in another folder that's part of the package. 

I'm telling PyInstaller to load those files specifically with --add-binary like this: `--add-binary 'PATH\\TO\\LIBRAY\\library.dll;.' `
I've tried using the spec file also to add the binaries, in addition I've tried using the PyInstaller options to load binaries from module aswell. 

Sorry for being a bit vague as I can't share the exact code. :)

**Reply by rokm** [reply]
> Sorry for being a bit vague as I can't share the exact code. :)

I suppose in that case, you could instrument [loader/pyimod04_ctypes.py](https://github.com/pyinstaller/pyinstaller/blob/v4.9/PyInstaller/loader/pyimod04_ctypes.py) to see if it is at least trying to load the correct library. And check if its dependencies are available in the frozen application.

**Reply by TheGrandJonsson** [reply]
I checked and it seems it's missing in the frozen application. I will try to build again later and see if I can dump the binaries at top level as you suggested earilier. Thanks!
